File: src/test/java/com/vjiki/music/controller/UserControllerIntegrationTest.java
package com.vjiki.music.controller;

import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.BDDMockito.given;
import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.user;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import java.time.OffsetDateTime;
import java.util.UUID;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.http.MediaType;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.web.servlet.MockMvc;

import com.vjiki.music.dto.UserResponse;
import com.vjiki.music.service.UserService;

@WebMvcTest(UserController.class)
@DisplayName("UserController integration tests")
class UserControllerIntegrationTest {

    @Autowired
    private MockMvc mockMvc;

    @MockitoBean
    private UserService userService;

    @Nested
    @DisplayName("GET /api/v1/users/{userId}")
    class GetUserById {

        @Test
        @DisplayName("should return user when found")
        void shouldReturnUserWhenFound() throws Exception {
            UUID userId = UUID.randomUUID();
            UserResponse userResponse = UserResponse.builder()
                    .id(userId)
                    .email("test@example.com")
                    .nickname("testuser")
                    .avatarUrl(null)
                    .accessLevel("USER")
                    .isActive(true)
                    .isVerified(false)
                    .lastLoginAt(null)
                    .createdAt(OffsetDateTime.now())
                    .build();

            given(userService.getUserById(userId)).willReturn(userResponse);

            mockMvc.perform(get("/api/v1/users/{userId}", userId).with(user("test").roles("USER")))
                    .andExpect(status().isOk())
                    .andExpect(content().contentType(MediaType.APPLICATION_JSON))
                    .andExpect(jsonPath("$.id").value(userId.toString()))
                    .andExpect(jsonPath("$.email").value("test@example.com"))
                    .andExpect(jsonPath("$.nickname").value("testuser"));
        }

        @Test
        @DisplayName("should fail the request when user not found")
        void shouldFailWhenUserNotFound() throws Exception {
            UUID userId = UUID.randomUUID();
            given(userService.getUserById(userId)).willThrow(new RuntimeException("User not found"));

            // The application declares no exception handler, so the failure surfaces as a servlet
            // level error instead of a rendered 5xx response body.
            assertThatThrownBy(() -> mockMvc.perform(
                    get("/api/v1/users/{userId}", userId).with(user("test").roles("USER"))))
                    .rootCause()
                    .isInstanceOf(RuntimeException.class)
                    .hasMessage("User not found");
        }
    }
}

File: src/test/kotlin/com/vjiki/music/controller/UserControllerIntegrationTest.kt
package com.vjiki.music.controller

import com.vjiki.music.dto.UserResponse
import com.vjiki.music.service.UserService
import com.fasterxml.jackson.databind.ObjectMapper
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.boot.test.context.TestConfiguration
import org.springframework.context.annotation.Bean
import org.springframework.http.MediaType
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*
import java.time.OffsetDateTime
import java.util.*

@WebMvcTest(UserController::class)
class UserControllerIntegrationTest(
    @Autowired val mockMvc: MockMvc,
    @Autowired val objectMapper: ObjectMapper
) : DescribeSpec({

    val userService = mockk<UserService>()

    describe("GET /api/v1/users/{userId}") {
        it("should return user when found") {
            val userId = UUID.randomUUID()
            val userResponse = UserResponse(
                id = userId,
                email = "test@example.com",
                nickname = "testuser",
                avatarUrl = null,
                accessLevel = "USER",
                isActive = true,
                isVerified = false,
                lastLoginAt = null,
                createdAt = OffsetDateTime.now()
            )

            every { userService.getUserById(userId) } returns userResponse

            mockMvc.perform(get("/api/v1/users/$userId"))
                .andExpect(status().isOk)
                .andExpect(content().contentType(MediaType.APPLICATION_JSON))
                .andExpect(jsonPath("$.id").value(userId.toString()))
                .andExpect(jsonPath("$.email").value("test@example.com"))
                .andExpect(jsonPath("$.nickname").value("testuser"))
        }

        it("should return 500 when user not found") {
            val userId = UUID.randomUUID()
            every { userService.getUserById(userId) } throws RuntimeException("User not found")

            mockMvc.perform(get("/api/v1/users/$userId"))
                .andExpect(status().is5xxServerError)
        }
    }
})

